File: exercism/ruby/bob/bob.rb
class Remark
  def initialize(phrase)
    @phrase = phrase.strip
  end

  def question?
    @phrase[-1] == '?'
  end

  def insult?
    @phrase.match(/[A-Z]/i) && @phrase.upcase == @phrase
  end

  def no_content?
    @phrase.scan(/\s/).join == @phrase
  end
end

module Bob
  def self.hey(phrase)
    @@remark = Remark.new(phrase)
    reply
  end

  def self.reply
    if @@remark.insult?
      'Whoa, chill out!'
    elsif @@remark.question?
      'Sure.'
    elsif @@remark.no_content?
      'Fine. Be that way!'
    else
      'Whatever.'
    end
  end
end

module BookKeeping
  VERSION = 1
end

File: exercism/javascript/bob/bob.js
function Bob() {}

Bob.prototype.hey = function hey(req) {
  var reqUpper = req.toUpperCase();
  var reqWhitoutBlanks = req.replace(/\s/g, '');
  var reqOnlyLetters = req.replace(/[^a-zA-Z]/g, '');
  var lastNonBlankChar = reqWhitoutBlanks[reqWhitoutBlanks.length - 1];
  var resp;

  var question = (lastNonBlankChar === '?');
  var shouting = (reqUpper === req && reqOnlyLetters !== '');
  var silence = (reqWhitoutBlanks === '');

  if (question && shouting) {
    resp = "Calm down, I know what I'm doing!";
  } else if (question) {
    resp = "Sure.";
  } else if (shouting) {
    resp = "Whoa, chill out!";
  } else if (silence) {
    resp = "Fine. Be that way!";
  } else {
    resp = "Whatever."
  }

  return resp;
}

module.exports = Bob;
